nemo/collections/nlp/models/dialogue/dialogue_gpt_classification_model.py

```diff
@@ -224,47 +267,72 @@ def predict_step(self, batch, batch_idx, dataloader_idx=None):
         # return self(batch)
         raise NotImplementedError()
 
+    def on_train_end(self):
+        if self.prompt_learning:
+            self.language_model.on_train_end()
+
+    def get_prompt_token_labels_for_megatron_gpt(self, input_ids, num_prompt_tokens):
+
+        prompt_token_labels = torch.full(
+            size=(input_ids.size(0), num_prompt_tokens),
+            fill_value=self.tokenizer.tokenizer.pad_token_id,
+            dtype=torch.long,
+        )
+
+        if self.prompt_learning:
+            prompt_token_labels.data = torch.LongTensor(
+                np.tile(np.array(self.language_model.pseudo_token_ids), (input_ids.size(0), 1))
+            )
+
+        prompt_token_labels = prompt_token_labels.to(input_ids.device)
+
+        return prompt_token_labels
+
+    def get_prompt_ids_for_megatron_gpt(self, input_ids):
+        if self.cfg.virtual_prompt_style == 'prompt-tuning' or not self.prompt_learning or self.trainer.testing:
+            prompt_ids = torch.tensor([0] * input_ids.size(0)).to(input_ids.device) if self.prompt_learning else None
+        else:
+            total_virtual_tokens = self.cfg.task_templates[0].total_virtual_tokens
+            init_text = self.cfg.prompt_tuning.new_prompt_init_text[0]
+            init_text_ids = self.tokenizer.text_to_ids(init_text)
+            init_text_ids = torch.tensor(init_text_ids).to(input_ids.device)
+            prompt_ids = init_text_ids.repeat(input_ids.size(0), 1)[:, :total_virtual_tokens]
+        return prompt_ids
+
     def forward(self, input_ids, attention_mask, labels, inference=True):
 
         if self.cfg.library == "huggingface":
             output = self.language_model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
             loss = output['loss']
 
+            b_logits = output['logits']
+            # Shift so that tokens < n predict n
+            shift_logits = b_logits[..., :-1, :].contiguous()
+            shift_labels = labels[..., 1:].contiguous()
+            loss_fct = torch.nn.CrossEntropyLoss(reduction='none')
```

Comment:
Why do you reinitialize the loss function during every forward call? The loss function should probably be a class attribute that gets initialized once.

Reply:
Good catch!